src/types/boxed.rs

```diff
@@ -0,0 +1,343 @@
+///
+/// Passing some immutable data between Rust and JavaScript.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # use std::path::{Path, PathBuf};
+/// fn create_path(mut cx: FunctionContext) -> JsResult<JsBox<PathBuf>> {
+///     let path = cx.argument::<JsString>(0)?.value(&mut cx);
+///     let path = Path::new(&path).to_path_buf();
+///
+///     Ok(cx.boxed(path))
+/// }
+///
+/// fn print_path(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+///     let path = cx.argument::<JsBox<PathBuf>>(0)?;
+///
+///     println!("{}", path.display());
+///
+///     Ok(cx.undefined())
+/// }
+/// ```
+///
+/// Passing a user defined struct wrapped in a `RefCell` for mutability. This
+/// pattern is useful for creating classes in JavaScript.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # use std::cell::RefCell;
+///
+/// type BoxedPerson = JsBox<RefCell<Person>>;
+///
+/// struct Person {
+///      name: String,
+/// }
+///
+/// impl Finalize for Person {}
+///
+/// impl Person {
+///     pub fn new(name: String) -> Self {
+///         Person { name }
+///     }
+/// 
+///     pub fn set_name(&mut self, name: String) {
+///         self.name = name;
+///     }
+/// 
+///     pub fn greet(&self) -> String {
+///         format!("Hello, {}!", self.name)
+///     }
+/// }
+/// 
+/// fn person_new(mut cx: FunctionContext) -> JsResult<BoxedPerson> {
+///     let name = cx.argument::<JsString>(0)?.value(&mut cx);
+///     let person = RefCell::new(Person::new(name));
+/// 
+///     Ok(cx.boxed(person))
+/// }
+/// 
+/// fn person_set_name(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+///     let person = cx.argument::<BoxedPerson>(0)?;
+///     let mut person = person.borrow_mut();
+///     let name = cx.argument::<JsString>(1)?.value(&mut cx);
+/// 
+///     person.set_name(name);
+/// 
+///     Ok(cx.undefined())
+/// }
+/// 
+/// fn person_greet(mut cx: FunctionContext) -> JsResult<JsString> {
+///     let person = cx.argument::<BoxedPerson>(0)?;
+///     let person = person.borrow();
+///     let greeting = person.greet();
+/// 
+///     Ok(cx.string(greeting))
+/// }
+pub struct JsBox<T: Send + 'static> {
+    local: raw::Local,
+    // `JsBox` cannot verify the lifetime. Store a raw pointer to force uses
+    // to be marked unsafe. In practice, it can be treated as `'static` but
+    // should only be exposed as part of a `Handle` tied to a `Context` lifetime.
+    internal: *const T,
+}
+
+// Custom `Clone` implementation since `T` might not be `Clone`
+impl<T: Send + 'static> Clone for JsBox<T> {
+    fn clone(&self) -> Self {
+        JsBox {
+            local: self.local,
+            internal: self.internal,
+        }
+    }
+}
+
+impl<T: Send + 'static> Copy for JsBox<T> {}
+
+impl<T: Send + 'static> Value for JsBox<T> { }
+
+impl<T: Send + 'static> Managed for JsBox<T> {
+    fn to_raw(self) -> raw::Local {
+        self.local
+    }
+
+    fn from_raw(env: Env, h: raw::Local) -> Self {
+        let v = unsafe {
```

Comment:
Maybe this would be a little easier to follow, and with slightly more specific error messages, if we unwrap the optional results at each step? Something like:

```rust
fn from_raw(env: Env, h: raw::Local) -> Self {
    let p: *const T = external::deref::<BoxAny>(env.to_raw(), h)
        .expect("JsBox failed to unwrap external");
    let r: &T = &*p;
    let unwrapped = r.downcast_ref()
        .expect("JsBox failed to downcast external");

    Self {
        local: h,
        unwrapped
    }
}
```